src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts

```diff
@@ -360,26 +375,51 @@ export class DataView implements IIndexPattern {
 
   /**
    * Add a runtime field - Appended to existing mapped field or a new field is
-   * created as appropriate
+   * created as appropriate. We can pass along with the fields a "format" definition or "customLabel" to avoid calling
+   * separately `setFieldCustomLabel()` and `setFieldFormat()`.
    * @param name Field name
-   * @param runtimeField Runtime field definition
+   * @param enhancedRuntimeField Runtime field definition
    */
-  addRuntimeField(name: string, runtimeField: RuntimeField) {
+  addRuntimeField(name: string, enhancedRuntimeField: EnhancedRuntimeField): DataViewField {
```

Comment:
The naming of `EnhancedRuntimeField` vs `RuntimeField` is a bit confusing to me. All of the functions don't use the `enhanced` prefix, but they take the enhanced version in as a parameter. Should this be `addEnhancedRuntimeField`? Can you add a regular `RuntimeField`? Should we rethink the names?  It also looks like maybe `RuntimeField` type doesn't need to be public anymore. The only place I found outside this plugin using  it is the field editor for `runtimeFieldValidator`, and the only usage is:

```
        const error = await runtimeFieldValidator({
          type: updatedField.type,
          script: updatedField.script,
        });
```

Maybe that can use an internal type and we only need to export publicly one type of "RuntimeField"?

Reply:
Let's take one step back to clarify the problem I am trying to solve:

The field attributes (`customLabel`) and the field formats are not stored on the `DataViewField`.   The former is stored in the `fieldAttrs` map and the later in `fieldFormatMap`. To set them we need to first create the fields then call 2 methods on the index pattern.

The `runtimeField` object which is saved on the `DataViewField` is simply to store the ES runtime field interface + a reference to an optional parent composite. 

```
interface RuntimeField extends ESRuntimeField {
  parentComposite?: string;
}
```

**This interface** is what is saved in each of the `DataViewField` fields and also inside our `runtimeFieldMap`. As [you can see here](https://github.com/elastic/kibana/pull/110226/files#diff-64a353672cfc941c11c93ce19983c50f5de7ae626fa93b35a402f5588ac11c71R394).

Great, so we do need this interface.

Now, I need another interface where the consumer can pass along additional data like format definition, custom label and popularity. And that's the `EnhancedRuntimeField`.

You could say, why not add a third parameter to the `addRuntimeField` method?

```js
addRuntimeField(name: string, runtimeField: RuntimeField, config?: { format?: SerializedFieldFormat, customLabel?: string; popularity?: number }) {
  ...
}
```

That could work... except that I still need a single interface for the runtime field + additional configuration **when adding a composite runtime field**.

```js
const myRuntimeComposite: {
  script: '<script that emits multiple values>',
  fields: {
    fieldOne: { ... }, // I want to pass format, customLabel and popularity along
    fieldTwo: { ... } // here too!
  }
};

indexPattern.addRuntimeComposite('myComposite', myRuntimeComposite);
```

We want to avoid the consumer to have to do this

```js
const compositeName: string = 'myComposite';
const compositeScript = { source: '<script that emit multiple values>' };

const subFields = {
  fieldOne: { type: 'keyword', customLabel: 'myLabel' },
  fieldTwo: { type: 'long', format: { ... } },
};

// note below we use the **RuntimeField** interface
const fieldsInComposite: Record<string, RuntimeField> = {};

Object.entries(subFields).forEach(([name, field]) => {
  fieldsInComposite[name] = { type: field.type };
  
  // we populate the different maps
  if (field.customLabel) {
    indexPattern.setFieldCustomLabel(name, field.customLabel);
  }
  if (field.format) {
    indexPattern.setFieldFormat(name, field.format);
  }
});

const runtimeComposite = {
  script: compositeScript,
  fields: fieldsInComposite, // those runtime fields are bare RuntimeField (no more enhanced)
}

indexPattern.addRuntimeComposite(compositeName, runtimeComposite);
```

To avoid all this boilerplate for the consumer I added the `EnhancedRuntimeField`.  

Note that the same boilerplate would need to be done when wanting to replace all runtime fields with new fields (`replaceAllRuntimeFields()`).

With all the above said, if you want to rename `EnhancedRuntimeField` with `RuntimeField` and rename the current `RuntimeField` to `DataViewFieldRuntimeDefinition` (and keep it internal to the `data` plugin) that could work.

Reply:
My above suggestion implies that we change the `FieldSpec.runtimeField` type to something like `FieldSpecRuntimeField`  (the current `RuntimeField` interface) ([here](https://github.com/elastic/kibana/blob/master/src/plugins/data/common/index_patterns/types.ts#L221))

Reply:
I need to take a closer look at this code but perhaps we could use a `FieldSpec` or perhaps define a `RuntimeFieldSpec`

Reply:
Before a user could do:

```
dv.addRuntimeField(name, field);
const sameField: RuntimeField = dv.getRuntimeField(name)'
```

Now you have
```
// Why is this returning a DataViewField type?
const field: DataViewField = dv.addRuntimeField(name, runtimeField);

// Here you are returning a different type, you've lost the extra fields (customLabel, etc).
const rtField: RuntimeField = dv.getRuntimeField(name);
```

I think the original was a better DX. As a consumer, I'd be surprised to get two different types.

From a consumers POV, is there any reason they wouldn't want the extra fields on the return type? Maybe it should be:

```
interface RuntimeField {
  ...
  customLabel: string;
  format: FieldFormat;
};

dv.addRuntimeField(name, field);
const field: RuntimeField = dv.getRuntimeField(name);
doSomthing(field.customLabel); 
```

Reply:
I'd prefer to keep `addRuntimeField` a simple method that only appends to the runtimeFieldMap. 

The field list will still need to be updated. Currently we generate the field list when a data view is generated and then modify it as `fieldAttrs` and similar are modified. We're making two sets of changes and its possible for things to get out of sync. We should just rebuild the field list whenever there's a modification. Shoot, `count` makes that a bad idea since its updated so frequently. Maybe we should do that for everything EXCEPT `count` updates. 

Then perhaps we have an additional function, `addRuntimeFieldWithAttrs` which is the same as `addRuntimeField` except it takes an additional argument which supplies the count, formatting, and custom name - the same values provided by `EnhancedRuntimeField` minus the class its extending. However, instead of only using it for runtime fields it should be 'core' to data view field types and used in their definition - IFieldType and FieldSpec.